### DISCUSSION TITLE
Why are stack traces enabled in CI (GitHub Actions)?

### DISCUSSION THREAD

**Discussion by jasonkuhrt** [oclif/core#1277] [Q&A] [answered]
I've been debugging without success why OClif is showing stack traces for errors in our CI pipeline. It breaks our snapshot tests that do not expect stack traces to be displayed.

I'm resorting to using pnpm patch with added internal logging to figure this out. My presumption is that the logging comes from the `handle` function https://github.com/oclif/core/blob/main/src/errors/handle.ts#L30. Perhaps the debug property is true, but if so, I don't know why it would be.

Here's an example test failure from a recent run:

![CleanShot 2024-12-20 at 13 38 13@2x](https://github.com/user-attachments/assets/76ac2662-02bc-4b43-af6c-fa801c8019ca)

**Comment by jasonkuhrt** [comment]
Turns out debug is true, now trying to understand _why_ that is.

![CleanShot 2024-12-20 at 13 52 23@2x](https://github.com/user-attachments/assets/4faa431b-71cf-422b-9ead-73333947b947)

**Reply by mdonnalley** [reply]
The `debug` setting is set here: https://github.com/oclif/core/blob/main/src/execute.ts#L58-L62

**Reply by jasonkuhrt** [reply]
Thanks, I'm aware. I've narrowed down my problem to around here:

![CleanShot 2024-12-20 at 14 56 15@2x](https://github.com/user-attachments/assets/d3162eab-0968-474c-9e40-d41815745a57)

It seems the code in the repo has changed since, not seeing `_debug` there https://github.com/oclif/core/blob/db95c989e9c5c3101c70c043efd8fcc99cc67b9f/src/config/config.ts.

**Reply by jasonkuhrt** [reply, marked as answer]
So, OClif sets debug true if `<BIN_NAME>_DEBUG=true` is in the environment. This was not clear to me. If that's documented, I missed it. Is it?

**Reply by mdonnalley** [reply]
What version of oclif/core are you using? The latest version no longer has that `_debug` method on `Config`

**Reply by jasonkuhrt** [reply]
We're currently on `@oclif/core@3.26.6`.

Does the new version not have the behaviour I discovered above?

**Reply by mdonnalley** [reply]
I think it's showing stack traces because of something on your end. How are you executing your CLI?

With a newly generated CLI, you can see that the stack trace only shows up if you use `bin/dev.js`

```
~/code/gh-discussion-1277 via ⬢ v20.17.0
❯ bin/run.js hello INVALID
 ›   Error: The following error occurred:
 ›     Missing required flag from
 ›   See more help with --help

~/code/gh-discussion-1277 via ⬢ v20.17.0
❯ bin/dev.js hello INVALID
Error: The following error occurred:
  Missing required flag from
See more help with --help
    at validateFlags (/Users/mdonnalley/code/gh-discussion-1277/node_modules/@oclif/core/lib/parser/validate.js:77:19)
    at async Object.parse (/Users/mdonnalley/code/gh-discussion-1277/node_modules/@oclif/core/lib/parser/index.js:20:5)
    at async Hello.parse (/Users/mdonnalley/code/gh-discussion-1277/node_modules/@oclif/core/lib/command.js:254:25)
    at async Hello.run (file:///Users/mdonnalley/code/gh-discussion-1277/src/commands/hello/index.ts:16:33)
    at async Hello._run (/Users/mdonnalley/code/gh-discussion-1277/node_modules/@oclif/core/lib/command.js:312:22)
    at async Config.runCommand (/Users/mdonnalley/code/gh-discussion-1277/node_modules/@oclif/core/lib/config/config.js:435:25)
    at async run (/Users/mdonnalley/code/gh-discussion-1277/node_modules/@oclif/core/lib/main.js:96:16)
    at async file:///Users/mdonnalley/code/gh-discussion-1277/bin/dev.js:6:1
```

If you remove the `settings.debug = true` line that happens inside of the `execute` function, then the stack trace isn't printed. So that tells me that you're likely using the dev script for your snapshot tests

**Reply by jasonkuhrt** [reply]
We're not using dev. But we did have HIVE_DEBUG envar set which I had not noticed.

**Reply by jasonkuhrt** [reply]
> I think it's showing stack traces because of something on your end.

Yes, the aforementioned `HIVE_DEBUG` envar.

> How are you executing your CLI?

```ts
const binPath = resolve(__dirname, '../../packages/libraries/cli/bin/run');
const cliDir = resolve(__dirname, '../../packages/libraries/cli');

export async function exec(cmd: string) {
  const result = await execaCommand(`${binPath} ${cmd}`, {
    shell: true,
    env: {
      OCLIF_CLI_CUSTOM_PATH: cliDir,
      NODE_OPTIONS: '--no-deprecation',
    },
  });

  if (result.failed) {
    throw new Error('CLI execution marked as "failed".', { cause: result.stderr });
  }

  return result.stdout;
}
```

> So that tells me that you're likely using the dev script for your snapshot tests

Right, that could have been, but not my case it turns out. If you look at the code I referenced before, at that version of OClif, there was at least then also logic to process any env of the pattern `<bin name>_DEBUG=true` which also will flip `settings.debug = true` deep in the code.
